test/acceptance/utils.ts

```diff
@@ -0,0 +1,14 @@
+import { CapturedResponse } from './../../src/models/frame';
+
+export function wait(time: number) {
+  return new Promise((r) => {
+    setTimeout(r, time);
+  });
+}
+
+export function getBody(response: CapturedResponse) {
```

Comment:
Is this util necessary? Why do we have this guard for `body`, but not `responseMeta`?

Reply:
responseMeta is required, but body is optional. Bodies are only available if the dev used waitForLoad = true, so trying to get a body when it doesn't exist should be an error case (devs won't try to access just to see if it's there).

```
export interface CapturedResponse {
  responseMeta: Network.Response;
  body?: string;
}
```

Reply:
I think we can provide safety here without having to add an extra utility. We can effectively have a conditional return type, based off of the parameter, by using overloaded param types.

For example: https://tinyurl.com/y96qpqmy

Alternatively, we can just go ahead and have two separate methods for `navigate`

bottom line is, it would be nicer to be able to eliminate possibility of error, rather than having possibility of runtime error.

Reply:
That's awesome.  I had considered either a dual return type (still with run-time guards), or bifurcating the method (with glue logic to share code). I had no idea TS had method signature overriding. This is the exact solution I need.

Reply:
Yea it's a nice feature. I believe it should be used somewhat sparingly though, as it can definitely make the function too complex. In most cases it's probably better to have two different functions, but I'm ok with using the conditional return type here.

Reply:
I think we should just drop the ability to return at nav complete time all together. Below is why:

--- TS method signature overriding ---
I tried it out, but it turns out to be too much of a hack to get working in this more complex case. The issue is that you literally have to pass true/false, not just a boolean variable. This work fine for the first call, where the user might literally call myMethod(true), but in my case the type we want to split on is three functions deep, which mean you'd have to start calling all the nested child functions like this:

if (waitForLoad) {
  childFunction(args, true);
else {
  childFunction (args, false);
} 

as well as have the 3 type signatures for each child function.

--- Split into two functions ---
The code isn't as elegant. I had to make a base function and three interfaces for the "wait=false" call, and another three interfaces & 30 lines of code to take the result obj apart, get bodies, and then jam it all back together.

--- With Either Approach ---
We already have 2 util functions in ember-li-sw which:
A) Are used in almost every test (both wait=true & wait=false tests)
B) Wrap navigate()
Since we now have two navigate functions, I would have to now split those two util functions into four util functions in order to maintain having only one type returned. You can imagine how this won't scale for the future if we add more util functions like this. I could use the "TS method signature overriding" trick to fork off into only one of the two nav functions and maintain only one instance of a util function, but at this point the code base is getting pretty complex.

All of this seem like way too much complexity in the code to support a "return after only nav complete" function that we don't need right now. For our test app (in both ember-li-sw and swae), turning all to waitForLoad doesn't significantly change the total test run time (delta sits within noise variation).

Reply:
My main issue with `getBody` is that the only value the util is providing is a runtime check, which could be a static compile time check.

> we now have two navigate functions

What are these two functions? `navigate` and ...?

> --- Split into two functions ---
> The code isn't as elegant. I had to make a base function and three interfaces for the ? 
"wait=false" call, and another three interfaces & 30 lines of code to take the result obj apart, get bodies, and then jam it all back together.

Is that all really necessary? Why can't we just rename current `navigate()` to `private loadPage()` and then create `load()` and `navigate()` where the former passes `waitForLoad=true`

> All of this seem like way too much complexity in the code to support a "return after only nav complete" function that we don't need right now. For our test app (in both ember-li-sw and swae), turning all to waitForLoad doesn't significantly change the total test run time (delta sits within noise variation).

So I don't yet buy that it's really that complicated to support both cases. Did you run your perf tests on CI tooling? If it really doesn't make any test perf difference, I'm fine dropping support for short circuiting waiting for the body.